python/cuml/metrics/cluster/silhouette_score_impl.pyx

```diff
@@ -0,0 +1,88 @@
+import cupy as cp
+import numpy as np
+
+from libc.stdint cimport uintptr_t
+
+from cuml.common import input_to_cuml_array
+from cuml.metrics.pairwise_distances import _determine_metric
+from cuml.raft.common.handle cimport handle_t
+from cuml.raft.common.handle import Handle
+from cuml.metrics.distance_type cimport DistanceType
+
+
+cdef extern from "cuml/metrics/metrics.hpp" namespace "ML::Metrics":
+    double silhouette_score(const handle_t &handle,
+                            double *y,
+                            int n_rows,
+                            int n_cols,
+                            int *labels,
+                            int n_labels,
+                            double *sil_scores,
+                            DistanceType metric) except +
+
+
+def _silhouette_coeff(
```

Comment:
Any particular reason to divide the cython and python code with an impl? For the most part we haven't done that, except for some cases where there is a cdef class.

Reply:
I went back and forth on this a bit, but I wanted to draw a sharp line between the C++ function `silhouette_score` and the Python function, which (for scikit-learn compatibility) has a different usage and API. My thought was that having these in separate files clearly indicates that there is some different "core" functionality that is being wrapped by the Python `silhouette_score`. If there is a separate impl file, there is less likely to be confusion about the difference behavior because people are more likely to look at both files alongside one another.

That being said, I was on the fence about that justification before, so I'm happy to bring these together if you think it would be better.

Reply:
Never mind; looking over some other algorithms, it's clear how this should be structured for consistency. I've updated accordingly.